testservices/identityservice/v3userpass.go

```diff
@@ -0,0 +1,198 @@
+		if ep.AdminURL != "" {
+			eps = append(eps, V3Endpoint{
+				Region:    ep.Region,
+				Interface: "admin",
+				URL:       ep.AdminURL,
+			})
+		}
+		if ep.InternalURL != "" {
+			eps = append(eps, V3Endpoint{
+				Region:    ep.Region,
+				Interface: "internal",
+				URL:       ep.InternalURL,
+			})
+		}
+		if ep.PublicURL != "" {
+			eps = append(eps, V3Endpoint{
+				Region:    ep.Region,
+				Interface: "public",
+				URL:       ep.PublicURL,
+			})
+		}
+	}
+	s.Endpoints = eps
+	u.services = append(u.services, s)
+}
+
+func (u *V3UserPass) ReturnFailure(w http.ResponseWriter, status int, message string) {
+	e := ErrorWrapper{
+		Error: ErrorResponse{
+			Message: message,
+			Code:    status,
+			Title:   http.StatusText(status),
+		},
+	}
+	if content, err := json.Marshal(e); err != nil {
+		w.Header().Set("Content-Length", fmt.Sprintf("%d", len(internalError)))
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write(internalError)
+	} else {
+		w.Header().Set("Content-Length", fmt.Sprintf("%d", len(content)))
+		w.WriteHeader(status)
+		w.Write(content)
+	}
+}
+
+func (u *V3UserPass) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var req V3UserPassRequest
+	// Testing against Canonistack, all responses are application/json, even failures
+	w.Header().Set("Content-Type", "application/json")
+	if r.Header.Get("Content-Type") != "application/json" {
+		u.ReturnFailure(w, http.StatusBadRequest, notJSON)
+		return
+	}
+	if content, err := ioutil.ReadAll(r.Body); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	} else {
+		if err := json.Unmarshal(content, &req); err != nil {
+			u.ReturnFailure(w, http.StatusBadRequest, notJSON)
+			return
+		}
+	}
+	userInfo, errmsg := u.authenticate(
+		req.Auth.Identity.Password.User.Name,
+		req.Auth.Identity.Password.User.Password,
+	)
+	if errmsg != "" {
+		u.ReturnFailure(w, http.StatusUnauthorized, errmsg)
+		return
+	}
+	res, err := u.generateV3TokenResponse(userInfo)
+	if err != nil {
+		u.ReturnFailure(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	if req.Auth.Scope.Project.Name != "" {
+		res.Project = &v3Project{
+			ID: u.addTenant(req.Auth.Scope.Project.Name),
+		}
+	}
+	content, err := json.Marshal(struct {
+		Token *V3TokenResponse `json:"token"`
+	}{
+		Token: res,
+	})
+	if err != nil {
+		u.ReturnFailure(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	w.Header().Set("X-Subject-Token", userInfo.Token)
+	w.WriteHeader(http.StatusCreated)
+	w.Write(content)
+}
+
+func (u *V3UserPass) generateV3TokenResponse(userInfo *UserInfo) (*V3TokenResponse, error) {
+	res := V3TokenResponse{}
+
+	res.Issued = time.Now()
+	res.Expires = res.Issued.Add(24 * time.Hour)
+	res.Methods = []string{"password"}
+	res.Catalog = u.services
+	res.User.ID = userInfo.Id
+
+	return &res, nil
+}
+
+// setupHTTP attaches all the needed handlers to provide the HTTP API.
+func (u *V3UserPass) SetupHTTP(mux *http.ServeMux) {
+	mux.Handle("/v3/auth/tokens", u)
```

Comment:
NOTE: Because V3UserPass handles a more specific URL paths, it needs to be registered before the older UserPass.